go/tools/builders/compilepkg.go

```diff
@@ -280,8 +270,18 @@ func compileArchive(
 	// constraints.
 	compilingWithCgo := haveCgo && cgoEnabled
 
+	// When coverage is set, source files will be modified during instrumentation. We should only run static analysis
+	// over original source files and not the modified ones.
+	// goSrcsNogo and cgoSrcsNogo are copies of the original source files for nogo to run static analysis.
+	goSrcsNogo := goSrcs
+	cgoSrcsNogo := cgoSrcs
+
 	// Instrument source files for coverage.
 	if coverMode != "" {
+		// deep copy original source files for nogo static analysis, avoid being modified by coverage.
+		goSrcsNogo = append([]string{}, goSrcs...)
+		cgoSrcsNogo = append([]string{}, cgoSrcs...)
```

Comment:
Nit: would it make more sense to always use a copy of the slice instead of only copying during coverage?

```suggestion
	// goSrcsNogo and cgoSrcsNogo are copies of the original source files for nogo to run static analysis.
	// deep copy original source files for nogo static analysis, avoid being modified by coverage.
	goSrcsNogo := append([]string{}, goSrcs...)
	cgoSrcsNogo := append([]string{}, cgoSrcs...)

	// Instrument source files for coverage.
	if coverMode != "" {
```

Reply:
Since the purpose for `goSrcsNogo` and `cgoSrcsNogo` is to retain the original source files un-modified by coverage, I'm only making deep copies when coverage is enabled, as deep copy is more expensive in terms of performance and memory usage. But I also agree the code will look cleaner if we always make a deep copy, wdyt?